### DISCUSSION TITLE
Switching between USB/UART control without needing two versions of firmware?

### DISCUSSION THREAD

**Discussion by fireup924** [grblHAL/RP2040#78] [Ideas]
I would like to have the user be able to switch between controlling over USB or over UART without having to have two versions of the firmware. Can I just turn on MPG and have all the commands sent through that port? Does MPG have all the same features as controlling over UART?

**Comment by terjeio** [comment]
> Can I just turn on MPG and have all the commands sent through that port?

Yes. But you will have to provide a means to switch to the MPG stream - either via physical signal or a real time command (available in the keypad plugin). A custom plugin may also be added to facilitate switching.

>  Does MPG have all the same features as controlling over UART?

Yes, when MPG mode is active the MPG stream is in full control.
